sw/host/opentitantool/src/command/transport.rs

```diff
@@ -32,6 +35,42 @@ impl CommandDispatch for TransportInit {
     }
 }
 
+/// Updates the firmware of the debugger/transport.  If no argument is given, opentitantool will
+/// fetch an appropriate firmware from a public archive (NOT IMPLEMENTED YET).  For instructions
+/// on how to build HyperDebug firmware locally, see
+/// https://docs.google.com/document/d/1ZEH7L5j9-wMw4tkW28-xt6JU5B6hTX0RdZD4h4OZzDo .
+#[derive(Debug, StructOpt)]
+pub struct TransportUpdateFirmware {
+    #[structopt(
+        short,
+        long,
+        help = "Local firmware file to use instead of official release"
+    )]
+    filename: Option<PathBuf>,
+}
+
+impl CommandDispatch for TransportUpdateFirmware {
+    fn run(
+        &self,
+        _context: &dyn Any,
+        transport: &TransportWrapper,
+    ) -> Result<Option<Box<dyn Annotate>>> {
+        let firmware = match self.filename.as_ref() {
+            Some(name) => Some(fs::read(name)?),
+            None => None,
+        };
+        let progress = app::progress_bar(50000 as u64);
```

Comment:
The comment for `UpdateFirmware` above says the progress range is `0 - 1000`, but I see 50000 here.

Reply:
Yes, this is a work in progress (hence the status as Draft PR), and I went back and forth here.  Setting a range of 0 - 1000 causes the UI to show "1 kilobyte" (or something like that), which is incorrect.  In the current iteration, where only the low level transport knows the size of the "official" firmware image, it would require additional parameters to the progress function to pass the size.  I want to settle on how the image is downloaded/passed, before fixing details of the progress bar.

Reply:
To address the issue of the caller not knowing the size of the image upfront, in case of the "default" image being requested, and to allow separate progress during erasing flash, and subsequent writing of flash memory, I have introduced a StagedProgressBar struct.  I think it is convenient to avoid implementing the small `pfunc` functions in each command.

Another thing, the Ti50-specific "resuce" bootloader protocol actually transfers only part of the provided image, which has until now caused the progress bar to end before it reached 50%.  I plan on switching the `bootload` command to also use this StagedProgressBar, allowing the protocol implementation to indicate exactly how many bytes are to be transferred, rather than the command assuming it is always the full image file.